src/main/kotlin/org/eclipse/uprotocol/cloudevent/factory/UCloudEvent.kt

```diff
@@ -318,5 +315,134 @@ interface UCloudEvent {
+
+                extractIntegerValueFromExtension("plevel", event).ifPresent { permissionLevel = it }
+
+            }
+            return uMessage {
+                attributes = msgAttributes
+                payload = msgPayload
+                source = sourceUUri
+            }
+        }
+
+        /**
+         * Get the Cloudevent from the UMessage<br>
+         * <b>Note: For now, only the value format of UPayload is supported in the SDK.If the UPayload has a reference, it
+         * needs to be copied to CloudEvent.</b>
+         * @param message The UMessage protobuf containing the data
+         * @return returns the cloud event
+         */
+        fun fromMessage(message: UMessage): CloudEvent {
+            val attributes: UAttributes = message.attributes
+            val builder: CloudEventBuilder =
+                CloudEventBuilder.v1().withId(LongUuidSerializer.instance().serialize(attributes.id))
+            builder.withType(getEventType(attributes.type))
+            builder.withSource(URI.create(LongUriSerializer.instance().serialize(message.source)))
+            val contentType = getContentTypeFromUPayloadFormat(message.payload.format)
+            if (contentType.isNotEmpty()) {
+                builder.withDataContentType(contentType)
+            }
+            // IMPORTANT: Currently, ONLY the VALUE format is supported in the SDK!
+            if (message.payload.hasValue()) {
+                builder.withData(message.payload.value.toByteArray())
+            }
+
+            if (attributes.hasTtl()) {
+                builder.withExtension("ttl", attributes.ttl)
+            }
+            if (attributes.hasToken()) {
+                builder.withExtension("token", attributes.token)
+            }
+
+            if (attributes.priorityValue > 0) {
+                builder.withExtension("priority", attributes.priority.name)
+            }
+
+            if (attributes.hasSink()) {
+                builder.withExtension("sink", URI.create(LongUriSerializer.instance().serialize(attributes.sink)))
+            }
+
+            if (attributes.hasCommstatus()) {
+                builder.withExtension("commstatus", attributes.commstatus)
+            }
+
+            if (attributes.hasReqid()) {
+                builder.withExtension("reqid", LongUuidSerializer.instance().serialize(attributes.reqid))
+            }
+
+            if (attributes.hasPermissionLevel()) {
+                builder.withExtension("plevel", attributes.permissionLevel)
+            }
+
+            return builder.build()
+
+        }
+
+        /**
+         * Retrieves the payload format enumeration based on the provided content type.
+         *
+         * @param contentType The content type string representing the format of the payload.
+         * @return The corresponding UPayloadFormat enumeration based on the content type.
+         */
+        private fun getUPayloadFormatFromContentType(contentType: String?): UPayloadFormat {
```

Comment:
does contentType need to be nullable here?

Reply:
Yes, it can be null as well. For null content type, we need to return UPayloadFormat.UPAYLOAD_FORMAT_PROTOBUF.